docs/Design.md

```diff
@@ -463,8 +463,110 @@ this is a change to the AST upstream.
+
+#include "external_project/LargeHeader.h"
+#include "MyProjectHeader.h"
+
+void doStuff(const ExternalClass *k) { /* impl */ }
+```
+
+Just by looking at forward-declaration with the name
+`something::ExternalClass`, an indexing process has
+no way to tell if the definition for `ExternalClass`
+lives in a file in the current package
+or in an external package.
+
+<details>
+<summary>
+Could we determine the package information based
+on the namespace, maybe with a user-supplied mapping
+from FQN regexes to package information?
+</summary>
+
+At first glance, it might seem that maybe we can
+determine the package information based on the namespace.
+For example, if the user could provide a mapping for
+namespaces (regexes) -> packages, and if we see 
+a forward declaration in (say) the `llvm::` namespace,
+and the mapping has `llvm::.* -> llvm@15.0.0`,
+we could use `llvm@15.0.0` as the package information
+for the forward declaration.
+
+However, there isn't always such a strong correspondence
+between namespaces and packages,
+especially for patterns relying on features like ADL
+and template specializations.
+
+ADL example: (from our own code!)
+
+```
+namespace llvm {
+llvm::raw_ostream &operator<<(llvm::raw_ostream &os, Escaped s) {..}
+}
+```
+
+Template specialization example: (via https://stackoverflow.com/a/8157967)
+
+```
+namespace std {
+template <> struct hash<MyType> {
+  size_t operator()(const MyType & x) const { .. }
+};
+}
+```
+
+In both of these cases, the corresponding function or template specialization
+lives in a different project than the project which "owns" the namespace.
+
+</details>
+
+So, in general, if we want proper navigation for forward declarations,
+we need to actually discover which file
+the definition lives in,
+and this means that we need to index every TU,
+even if the main file does not belong to the current package.
+
+This raises a further problem, what do we do when the definition
+is not part of the indexing process at all?
+There are 4 possible cases here:
+1. Third-party binary-only libraries: In this case, there's not
+   really much that can be done, since a definition isn't available.
+2. First-party libraries built separately: In this case, we can
+   potentially consume the SCIP index generated from indexing the
+   first-party library, to identify which forward declaration
+   belongs to which library.
+3. Platform headers (e.g. `unistd.h`): These can depend on standard
+   library headers. TODO: How should we handle this case?
+4. Standard library headers (e.g. `iostream`): We can double-check
+   if standard library headers include any platform headers.
+   If not, we can make the simplifying assumption that any
+   forward declarations in standard library headers belong
+   to the corresponding synthetic standard library package
+   (e.g. `libc++@15.0.0`).
+
+For simplicity, an initial implementation
+can put fake `SymbolInformation` values
+in the `external_symbols` list
```

Comment:
I like this direction, for this reason: Those symbols mean what the linker says they mean. (Including the dynamic linker!) Until then, they're externs. So I think the question is, how are we going to handle information about linkage?

In a company with one component used in multiple products, we could easily have cases where one symbol has multiple definitions—one where it is linked one way, one where it is linked another way. For example if I'm reading some random bit of graphics code it might be calling ANGLE or it might be calling the OpenGL driver; or I'm using one JVM or another; or I'm using one version of Qt or another or whatever.

For people who can invest effort, it would be cool to have configurations that you could toggle. On Chromium I often wanted to filter by "I'm interested in the Android build configuration rn" or "I'm interested in Mac|Windows|Linux rn". But some thought needs to go into how that would work.

Reply:
Yes, we've discussed this before on the team -- an index is really a snapshot of a _build_ (= a project + a specific build configuration), not a _project_ by itself. Unfortunately, we don't really have a notion of a build anywhere, it would need to be in the backend, in the UI etc.

One simpler short-term thing would be split out some of the existing code for index merging in the monorepo into a separate subcommand in the SCIP CLI. That way, if you generated different indexes for different configurations (in any language), you could merge them ahead-of-time and upload the merged index. The backend and UI wouldn't need to be build configuration aware for this.

But yeah, the general solution requires a lot more work throughout the entire stack, and we haven't really scoped it out yet.